File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        
        for char in s:
            if char in mapping.values():  # If it's an opening bracket
                stack.append(char)
            else:  # It's a closing bracket
                if not stack or stack[-1] != mapping[char]:
                    return False
                stack.pop()
        
        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        unordered_map<char, char> mp = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };

        for (char c : s) {
            // If closing bracket
            if (mp.count(c)) {
                // Check if stack is empty or mismatch
                if (st.empty() || st.top() != mp[c])
                    return false;
                st.pop();
            } else {
                // Opening bracket
                st.push(c);
            }
        }

        return st.empty();
    }
};